cardano-cli/src/Cardano/CLI/Read.hs

```diff
@@ -1086,13 +1099,38 @@ instance Error CostModelsError where
         ]
 
 readCostModels
-  :: File L.CostModels In
+  :: ()
+  => ShelleyBasedEra era
+  -> File L.CostModels In
   -> ExceptT CostModelsError IO L.CostModels
-readCostModels (File fp) = do
+readCostModels sbe (File fp) = do
   bytes <- handleIOExceptT (CostModelsErrorReadFile . FileIOError fp) $ LBS.readFile fp
   costModels <- firstExceptT (CostModelsErrorJSONDecode fp) . except $ Aeson.eitherDecode bytes
   when (null $ fromAlonzoCostModels costModels) $ throwE $ CostModelsErrorEmpty fp
+  forM_ (allValues @L.Language) $ checkCostModelSize costModels
   return costModels
+ where
+  checkCostModelSize :: L.CostModels -> L.Language -> ExceptT CostModelsError IO ()
+  checkCostModelSize models lang =
+    case Map.lookup lang (L.costModelsValid models) of
+      Nothing -> pure ()
+      Just (model :: L.CostModel) -> do
+        let actual = Map.size (L.costModelToMap model)
+            expected = languageToParameterCount lang
+        unless (expected == actual) $ throwE $ CostModelsErrorWrongSize expected actual
+        return ()
+  allValues :: forall a. (Bounded a, Enum a) => [a]
+  allValues = [minBound :: a .. maxBound]
+  languageToParameterCount :: L.Language -> Int
+  languageToParameterCount = \case
+    L.PlutusV1 -> length $ allValues @PlutusV1.ParamName -- 166
+    L.PlutusV2 ->
+      let nbParamNames = length $ allValues @PlutusV2.ParamName in -- 185
+      caseShelleyToBabbageOrConwayEraOnwards
+        (const $ nbParamNames - 10) -- Ten parameters were added to V2 in Conway, need to remove them here
```

Comment:
I'm not so sure about hardcoding that number 10. We should account for the fact that the number of params can change in the future, however it's unlikely. That's why there's jumping over many hoops in `Cardano.Api.Genesis` and `Test.Cardano.Api.Genesis`, so we will get a clear signal when things unexpectedly change.